File: packages/catalog-schemas/src/components/calc-editor-common.ts
import {z} from 'zod';
import {formFieldBaseSchema} from './form-card';

/**
 * Общие куски контракта РАСЧЁТНЫХ редакторов (КЕО, инсоляция): вид источника
 * значения, описатель поля, readonly-строка условий и submit.
 *
 * Почему свой словарь источников, а не констракшновый
 * (`constructionsFieldSourceSchema`, четыре вида): у расчётных модулей есть
 * значения, ПОСЧИТАННЫЕ агентом (отметка центра окна по этажу, худший день
 * периода) и ДОПУЩЕНИЯ (отражение фасада ρ_ф 0,5) — метка обязана их
 * различать. Расширять общий словарь нельзя: подписи и оформление
 * ConstructionsEditor/LiftEditor завязаны ровно на четыре вида, и правка
 * задела бы их капабилити.
 */
export const calcFieldSourceKindSchema = z.enum([
  'project',
  'question',
  'suggested',
  'calculated',
  'assumption',
]);

export const calcFieldSourceSchema = z
  .object({
    source: calcFieldSourceKindSchema,
    // Человеческое обоснование одной строкой («худший день периода — 22 апреля»).
    note: z.string().min(1).max(200).optional(),
  })
  .strict();

/**
 * Источники значений одного экрана (помещения, расчётной точки, здания): имя
 * поля → источник. Параллельный блок, а не обёртка над значениями: payload
 * submit'а от провенанса не меняется, наружу источники не уходят.
 */
export const calcFieldSourcesSchema = z.record(z.string(), calcFieldSourceSchema);

/**
 * Раскрытие поля по значению поля-триггера того же экрана (ветка затенения
 * схемы N1). Один механизм на компонент, без вложенных условий.
 */
export const calcRevealBySchema = z
  .object({
    field: z.string().min(1).max(80),
    values: z.array(z.string().min(1).max(120)).min(1),
  })
  .strict();

/**
 * Поле расчётного редактора — базовое поле FormCard плюс аддитивные ключи.
 * Новой параллельной модели поля в каталоге не заводим (прецедент LiftEditor).
 */
export const calcEditorFieldSchema = formFieldBaseSchema
  .extend({
    // `lookup` расчётным редакторам не нужен: справочники живут у агента.
    type: z.enum(['text', 'number', 'select', 'boolean']),
    hint: z.string().min(1).max(200).optional(),
    // Подпись поля в строке-сводке свёрнутой секции; без неё берётся `name`.
    shortLabel: z.string().min(1).max(40).optional(),
    // Пояснение к варианту списка («не указана — принята северная»); в submit
    // уходит само значение, не текст.
    optionNotes: z.record(z.string(), z.string()).optional(),
    revealBy: calcRevealBySchema.optional(),
    // Предупреждение, зависящее от значения САМОГО поля: «нет» у затенения —
    // подпись о принятом открытом горизонте. Текст приходит от агента.
    valueWarnings: z.record(z.string(), z.string()).optional(),
    // Границы предупреждающей проверки числа; нарушение не блокирует submit.
    min: z.number().optional(),
    max: z.number().optional(),
  })
  .strict();

/**
 * Строка блока «Условия». Значение приходит готовой строкой; с `type` строка
 * становится КОНТРОЛОМ и правится на месте, а правленое значение уходит в
 * submit тем же ключом `name`.
 *
 * Почему правится вообще: город (регион строительства) агент берёт из проекта,
 * но проект бывает не тот — «изменить только для расчёта» на макете и означает,
 * что правка меняет расчёт, а не данные проекта. Без `type` строка остаётся
 * readonly — так приходят ВЫВЕДЕННЫЕ значения (норматив e_н, методика): их
 * правка не имеет смысла, их пересчитывает агент.
 */
export const calcConditionSchema = z
  .object({
    name: z.string().min(1).max(80),
    label: z.string().min(1).max(120),
    value: z.string().min(1).max(200),
    note: z.string().min(1).max(200).optional(),
    // Без `type` — readonly-строка (прежнее поведение, ключ аддитивный).
    type: z.enum(['text', 'select', 'lookup']).optional(),
    options: z.array(z.string().min(1).max(200)).min(1).optional(),
    // Справочник для `lookup`: город строительства выбирается из справочника
    // СП 131, а не набирается руками — опечатка в городе меняет весь расчёт.
    referenceId: z.string().min(1).max(80).optional(),
    // Откуда значение: подпись под контролом и слагаемое счётчика футера.
    source: calcFieldSourceSchema.optional(),
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.type === 'select' && (value.options?.length ?? 0) === 0) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'condition of type "select" requires "options"',
        path: ['options'],
      });
    }

    if (value.type === 'lookup' && value.referenceId === undefined) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'condition of type "lookup" requires "referenceId"',
        path: ['referenceId'],
      });
    }
  });

/** Единственный action расчётного редактора — «Рассчитать». */
export const calcSubmitSchema = z
  .object({
    name: z.string().min(1).max(120),
    label: z.string().min(1).max(80),
  })
  .strict();

export type CalcFieldSourceKind = z.infer<typeof calcFieldSourceKindSchema>;
export type CalcFieldSource = z.infer<typeof calcFieldSourceSchema>;
export type CalcFieldSources = z.infer<typeof calcFieldSourcesSchema>;
export type CalcRevealBy = z.infer<typeof calcRevealBySchema>;
export type CalcEditorField = z.infer<typeof calcEditorFieldSchema>;
export type CalcCondition = z.infer<typeof calcConditionSchema>;
export type CalcSubmit = z.infer<typeof calcSubmitSchema>;

File: packages/catalog-python/src/ai37_a2ui_catalog/models/calc_editor_common.py
from __future__ import annotations

from typing import Annotated, Literal

from pydantic import Field, model_validator

from .form_card import FormFieldLookupValue, LookupSuggestMode
from .shared import StrictModel

# Вид источника значения РАСЧЁТНЫХ редакторов (КЕО, инсоляция). Шире
# констракшнового словаря: есть значения, посчитанные агентом (`calculated`), и
# допущения (`assumption`). Общий словарь CE/LE расширять нельзя — их подписи
# завязаны ровно на четыре вида.
CalcFieldSourceKind = Literal["project", "question", "suggested", "calculated", "assumption"]

# Тип поля расчётного редактора: `lookup` не нужен — справочники живут у агента.
CalcEditorFieldType = Literal["text", "number", "select", "boolean"]

# Тип правимого условия. Без него строка условий readonly — прежнее поведение.
CalcConditionType = Literal["text", "select", "lookup"]


class CalcFieldSource(StrictModel):
    source: CalcFieldSourceKind
    # Человеческое обоснование одной строкой.
    note: str = Field(default=None, min_length=1, max_length=200)


class CalcRevealBy(StrictModel):
    """Раскрытие поля по значению поля-триггера того же экрана."""

    field: str = Field(min_length=1, max_length=80)
    values: list[Annotated[str, Field(min_length=1, max_length=120)]] = Field(min_length=1)


class CalcEditorField(StrictModel):
    """Поле расчётного редактора: базовое поле FormCard плюс аддитивные ключи."""

    name: str = Field(min_length=1, max_length=80)
    label: str = Field(min_length=1, max_length=120)
    type: CalcEditorFieldType
    required: bool = None
    options: list[str] = Field(default=None, min_length=1)
    referenceId: str = Field(default=None, min_length=1, max_length=80)
    minChars: int = Field(default=None, ge=1, le=10)
    suggestMode: LookupSuggestMode = None
    placeholder: str = Field(default=None, min_length=1, max_length=120)
    defaultValue: str | int | float | bool | FormFieldLookupValue = None
    hint: str = Field(default=None, min_length=1, max_length=200)
    # Подпись поля в строке-сводке свёрнутой секции.
    shortLabel: str = Field(default=None, min_length=1, max_length=40)
    # Пояснение к варианту списка; в submit уходит само значение.
    optionNotes: dict[str, str] = None
    revealBy: CalcRevealBy = None
    # Предупреждение по значению самого поля («нет» у затенения).
    valueWarnings: dict[str, str] = None
    # Границы предупреждающей проверки числа; submit не блокируют.
    min: float = None
    max: float = None


class CalcCondition(StrictModel):
    """Строка блока «Условия».

    Значение приходит готовой строкой; с `type` строка становится контролом и
    правится на месте, а правленое значение уходит в submit тем же ключом
    `name`. Город (регион строительства) агент берёт из проекта, но проект
    бывает не тот — правка меняет расчёт, а не данные проекта. Без `type`
    строка остаётся readonly: так приходят выведенные значения (норматив,
    методика), их пересчитывает агент.

    Пустое `value` — отсутствие ответа, и допустимо оно только у правимого
    условия: иначе агент изобретает заглушку («—»), а она уезжает в submit как
    настоящее значение. Выведенное условие пустым не бывает — не вычислилось,
    значит строку не присылать вовсе.
    """

    name: str = Field(min_length=1, max_length=80)
    label: str = Field(min_length=1, max_length=120)
    value: str = Field(max_length=200)
    note: str = Field(default=None, min_length=1, max_length=200)
    type: CalcConditionType = Field(default=None)
    options: list[Annotated[str, Field(min_length=1, max_length=200)]] = Field(
        default=None, min_length=1
    )
    referenceId: str = Field(default=None, min_length=1, max_length=80)
    source: CalcFieldSource = Field(default=None)

    @model_validator(mode="after")
    def validate_control(self) -> "CalcCondition":
        if self.value == "" and self.type is None:
            raise ValueError('a non-empty "value" is required when "type" is absent')
        if self.type == "select" and not self.options:
            raise ValueError('options are required when type is "select"')
        if self.type == "lookup" and not isinstance(self.referenceId, str):
            raise ValueError('referenceId is required when type is "lookup"')
        return self


class CalcSubmit(StrictModel):
    """Единственный action расчётного редактора — «Рассчитать»."""

    name: str = Field(min_length=1, max_length=120)
    label: str = Field(min_length=1, max_length=80)
